pep-0688.rst

```diff
@@ -122,26 +122,39 @@ object, for example by the ``memoryview()`` constructor.
 It corresponds to the ``bf_getbuffer`` C slot.
 The Python signature for this method is
 ``def __buffer__(self, flags: int, /) -> memoryview: ...``. The method
-must return a ``memoryview`` object. If the method is called from C
-code, the interpreter extracts the underlying ``Py_buffer`` from the
-``memoryview`` and returns it to the C caller. Similarly, if the
-``__buffer__`` method is called on an instance of a C class that
+must return a ``memoryview`` object. If the ``bf_getbuffer`` slot
+is invoked on a Python class with a ``__buffer__`` method,
+the interpreter extracts the underlying ``Py_buffer`` from the
+``memoryview`` returned by the method
+and returns it to the C caller. Similarly, if Python code calls the
+``__buffer__`` method on an instance of a C class that
 implements ``bf_getbuffer``, the returned buffer is wrapped in a
 ``memoryview`` for consumption by Python code.
 
-The ``__release_buffer__`` method is called when a caller no
+The ``__release_buffer__`` method should be called when a caller no
 longer needs the buffer returned by ``__buffer__``. It corresponds to the
 ``bf_releasebuffer`` C slot. This is an
 optional part of the buffer protocol.
 The Python signature for this method is
 ``def __release_buffer__(self, buffer: memoryview, /) -> None: ...``.
 The buffer to be released is wrapped in a ``memoryview``. When this
 method is invoked through CPython's buffer API (for example, through
-``memoryview.release``), the passed ``memoryview`` is the same object
+calling ``memoryview.release`` on a ``memoryview`` returned by
+``__buffer__``), the passed ``memoryview`` is the same object
 as was returned by ``__buffer__``. It is
 also possible to call ``__release_buffer__`` on a C class that
 implements ``bf_releasebuffer``.
 
+If ``__release_buffer__`` exists on an object,
+Python code that calls ``__buffer__`` directly on the object must
+call ``__release_buffer__`` on the same object when it is done
+with the buffer. If not, resources used by the buffer may be
+leaked until the garbage collector reclaims the ``memoryview``
+returned by ``__buffer__``. Similarly, it is a programming error
+to call ``__release_buffer__`` without a previous call to
+``__buffer__``. The interpreter will ensure that such errors
+do not lead to memory leaks or memory safety violations.
```

Comment:
The intent sounds good, but the wording is too strong. For example, if `__buffer__` acquires a lock, that's a resource that won't be released when the `memoryview` is garbage-collected. Or storing something in a global dict will lead to a resource (and thus memory) leak.

How about this?
```suggestion
If ``__release_buffer__`` exists on an object,
Python code that calls ``__buffer__`` directly on the object must
call ``__release_buffer__`` on the same object when it is done
with the buffer.  Similarly, it is a programming error
to call ``__release_buffer__`` without a previous call to
``__buffer__``.
Both errors are similar in severity to calling ``__enter__``
without ``__exit__``. The interpreter will ensure that misuse
of the Python API will not break invariants at the C level -- for
example, it will not cause memory safety violations.
```